### PR TITLE
WGS T/N Workflow Launch

### LINKED ISSUE #218: [Launch Pad v2] Allow user to select T/N libraries and launch WGS T/N workflow
**Context:**

- See story point -- https://trello.com/c/ONt9UvRq/1263-gds-cleanup-add-portal-capability-to-rerun-samples

> We will need the ability to select two libraries identifiers (one tumor, one normal) and request a launch of our T/N WGS workflow on ICA.

**Actions:**

- Implement it

### PR DESCRIPTION
**Pull request by williamputraintan**
- Added WGS T/N Launch from UI via AWS Lambda SDK
- Refactoring on Launch Page Component

Some ScreenShots:
Screenshot 2023-02-21 at 13 00 07
CONFIRMATION
Screenshot 2023-02-21 at 13 03 18
Screenshot 2023-02-21 at 13 02 21

Fix #218 

Link API PR: https://github.com/umccr/data-portal-apis/pull/554
Link Infra PR: https://github.com/umccr/infrastructure/pull/282

### CODE REVIEW

**Comment by victorskl** [comment]
I will review this asap tomorrow.

**Review by victorskl — COMMENTED** [review]
As discussed; I was a bit miss-leading you on UX flow in [this comment](https://github.com/umccr/data-portal-client/issues/218#issuecomment-1430624800). My fault. Sorry, Will.

For the first cut; to simplify thing, we should make use of [pairing endpoint](https://github.com/umccr/data-portal-apis/blob/dev/docs/ENDPOINTS.md#pairing-endpoint) by subject.

If we target our audience to end user (i.e. no idea about Metadata relationship such as topup; all possible combination of FASTQ pairs i.e. Library <> Run <> Lane; etc... then it will be more counterproductive to offer selection criteria validation at frontend.

So, let us start with safer approach by utilising pairing endpoint. Then, we shall expand the use case from there in future iteration.

**Comment by victorskl** [comment]
These are few cases that you can use it for dev/testing purpose.

```
curl -s -X POST -d '["SBJ02863"]' -H "Content-Type: application/json" -H "Authorization: Bearer $PORTAL_TOKEN" "https://api.portal.prod.umccr.org/pairing/by_subjects" | jq

curl -s -X POST -d '["SBJ02899"]' -H "Content-Type: application/json" -H "Authorization: Bearer $PORTAL_TOKEN" "https://api.portal.prod.umccr.org/pairing/by_subjects" | jq

curl -s -X POST -d '["SBJ02686"]' -H "Content-Type: application/json" -H "Authorization: Bearer $PORTAL_TOKEN" "https://api.portal.prod.umccr.org/pairing/by_subjects" | jq

curl -s -X POST -d '["SBJ02096"]' -H "Content-Type: application/json" -H "Authorization: Bearer $PORTAL_TOKEN" "https://api.portal.prod.umccr.org/pairing/by_subjects" | jq
```

File: src/api/fastq.ts [hunk 1]
```diff
@@ -0,0 +1,34 @@
+import { API } from '@aws-amplify/api';
```
Comment (victorskl) [inline src/api/fastq.ts:1]: Having said; could please keep this. We might need it in future...

Comment (williamputraintan) [reply src/api/fastq.ts:1]: Yuppp 🫡

File: src/api/metadata.ts [hunk 2]
```diff
@@ -24,10 +24,15 @@ export type MetadataRow = {
...
+type usePortalMetadataAPIProps = {
+  additionalPath?: string;
+  apiConfig: Record<string, any>;
+};
+export function usePortalMetadataAPI({ additionalPath, apiConfig }: usePortalMetadataAPIProps) {
   return useQuery(
     ['portal-metadata', apiConfig],
-    async () => await API.get('portal', `/metadata/`, apiConfig),
+    async (): Promise<MetadataApiRes> =>
+      await API.get('portal', `/metadata/${additionalPath}`, apiConfig),
```
Comment (victorskl) [inline src/api/metadata.ts:35]: Looks like it breaks `MetadataHomeCard`, `MetadataTable`, pages... The `additionalPath` is optional or, it is not required? The call resolves to `undefined`.

```
http://localhost:8000/metadata/undefined?...
```

File: src/config.ts [hunk 3]
```diff
@@ -3,6 +3,8 @@ const REGION = import.meta.env.VITE_REGION;
 const IS_LOCAL = STAGE === 'localhost';
 const OAUTH_DOMAIN = `${import.meta.env.VITE_OAUTH_DOMAIN}.auth.${REGION}.amazoncognito.com`;
 
+const LAMBDA_PREFIX = `data-portal-api-${STAGE == 'prod' ? 'prod' : 'dev'}-`;
```
Comment (victorskl) [inline src/config.ts:6]: Ah. We also have `stg`... :)

File: src/api/pairing.ts [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+  output_file_prefix: string;
+  output_directory: string;
+  fastq_list_rows: FastqRow[];
+  tumor_fastq_list_rows: FastqRow[];
+};
+
+type UsePortalSubjectParingAPIProps = {
+  apiConfig: Record<string, any>;
+};
+export function usePortalSubjectParingAPI({ apiConfig }: UsePortalSubjectParingAPIProps) {
```
Comment (victorskl) [inline src/api/pairing.ts:33]: Nice!

File: src/containers/home/MetadataCard/index.tsx [hunk 5]
```diff
@@ -8,7 +8,7 @@ import { usePortalMetadataAPI } from '../../../api/metadata';
 function MetadataHomeCard() {
   const { toastShow } = useToastContext();
 
-  const { isFetching, isLoading, isError, data } = usePortalMetadataAPI({});
+  const { isFetching, isLoading, isError, data } = usePortalMetadataAPI({ apiConfig: {} });
```
Comment (victorskl) [inline src/containers/home/MetadataCard/index.tsx:11]: ditto missing first args `additionalPath`; it breaks API call.

File: src/containers/metadata/MetadataTable/index.tsx [hunk 6]
```diff
@@ -57,8 +57,10 @@ function MetadataTable() {
   type ObjKeyType = { [key: string]: string | number };
   let metadataDataList: ObjKeyType[] = [];
   const { isFetching, isLoading, isError, data } = usePortalMetadataAPI({
-    queryStringParameters: {
-      ...apiQueryParameter,
+    apiConfig: {
```
Comment (victorskl) [inline src/containers/metadata/MetadataTable/index.tsx:60]: ditto `additionalPath`

File: src/containers/subjects/SubjectWGSTNLaunch/index.tsx [hunk 7]
```diff
@@ -0,0 +1,311 @@
...
+
+  const pairingOptions = usePortalSubjectParingAPI({
+    apiConfig: {
+      header: { 'Content-Type': 'application/json' },
+      body: [subjectId],
+    },
+  });
+
+  const metadataUseQueryRes = usePortalMetadataAPI({
+    additionalPath: '?workflow=clinical&workflow=research&phenotype=tumor&phenotype=normal',
```
Comment (victorskl) [inline src/containers/subjects/SubjectWGSTNLaunch/index.tsx:54]: Ah. `additionalPath` is used here. Perhaps make it optional in `usePortalMetadataAPI(...)` I reckon. I leave it to you!

**Review by victorskl — APPROVED** [review]
Jolly good, Will! Nicely done. Minor comments to attend. Once those fix up and ready, please feel free to merge.

File: src/api/metadata.ts (hunk 2 above)
Comment (williamputraintan) [reply src/api/metadata.ts:35]: Arghh! Fixing it!

File: src/config.ts (hunk 3 above)
Comment (williamputraintan) [reply src/config.ts:6]: Ahhh forgot about that, updating...

File: src/api/metadata.ts (hunk 2 above)
Comment (williamputraintan) [reply src/api/metadata.ts:38]: Setting default value for empty string when it is missing. This should work

File: src/containers/subjects/SubjectWGSTNLaunch/index.tsx [hunk 8]
```diff
@@ -0,0 +1,311 @@
...
+  }
+
+  return (
+    <div>
+      <div className='text-2xl font-medium mb-4'>
+        {subjectId} - Whole-Genome Sequencing Tumor-Normal (WGS T/N) Launch
+      </div>
+      <h5 className='mt-0'>Description</h5>
+      <div>
+        {`This page should launch should able to launch WGS T/N workflow for Subject Id "${subjectId}". `}
```
Comment (pdiakumis) [inline src/containers/subjects/SubjectWGSTNLaunch/index.tsx:133]:
```suggestion
        {`This page should be able to launch the WGS T/N workflow for SubjectId "${subjectId}". `}
```

File: src/containers/subjects/SubjectWGSTNLaunch/index.tsx [hunk 9]
```diff
@@ -0,0 +1,311 @@
...
+        {`Please select relevant libraries for tumor and normal as well as other inputs if available. `}
+        {`These inputs will construct payload and invoke lambda describe at `}
```
Comment (pdiakumis) [inline src/containers/subjects/SubjectWGSTNLaunch/index.tsx:135]:
```suggestion
        {`These inputs will construct the payload and invoke the lambda described at `}
```

**Review by victorskl — APPROVED** [review]

File: src/containers/subjects/SubjectWGSTNLaunch/index.tsx [hunk 10]
```diff
@@ -0,0 +1,311 @@
...
+        <a
+          target={`_blank`}
+          href='https://github.com/umccr/data-portal-apis/blob/dev/docs/pipeline/automation/tumor_normal.md'>
+          data-portal-pipeline-t/n-docs
+        </a>
+        {`.`}
+      </div>
+
+      <h5>Lab Metadata Table</h5>
+      <div className='mb-3'>{`This is just an additional metadata table to help you to select relevant FASTQ.`}</div>
```
Comment (pdiakumis) [inline src/containers/subjects/SubjectWGSTNLaunch/index.tsx:145]:
```suggestion
      <div className='mb-3'>{`This is just an additional metadata table to help you select relevant FASTQs.`}</div>
```

**Review by pdiakumis — APPROVED** [review]
Just some minor typos that caught my eye in the screenshots @williamputraintan, looks good otherwise!

File: src/containers/subjects/SubjectWGSTNLaunch/index.tsx [hunk 11]
```diff
@@ -0,0 +1,311 @@
...
+  // SUCCESS component
+  if (workflowTriggerRes.isSuccess) {
+    return (
+      <div className='mt-3 text-center'>
+        <Button
+          icon='pi pi-check'
+          className='p-button-rounded p-button-success bg-green-700 cursor-auto'
+          aria-label='Cancel'
+        />
+        <div className='mt-3'>{`Successfully launch WGS T/N workflow! Check Slack for updates.`}</div>
```
Comment (pdiakumis) [inline src/containers/subjects/SubjectWGSTNLaunch/index.tsx:120]:
```suggestion
        <div className='mt-3'>{`Successfully launched WGS T/N workflow! Check Slack for updates.`}</div>
```

File: src/containers/subjects/SubjectWGSTNLaunch/index.tsx [hunk 12]
```diff
@@ -0,0 +1,311 @@
...
+  }
+  if (workflowTriggerRes.isError) {
+    return (
+      <div className='mt-3 text-center'>
+        <Button
+          icon='pi pi-times'
+          className='p-button-rounded p-button-danger bg-red-500 cursor-auto'
+          aria-label='Cancel'
+        />
+        <div className='mt-3'>{`Something went wrong on launching WGS T/N workflow!`}</div>
```
Comment (pdiakumis) [inline src/containers/subjects/SubjectWGSTNLaunch/index.tsx:91]:
```suggestion
        <div className='mt-3'>{`Something went wrong upon launching WGS T/N workflow!`}</div>
```

**Comment by williamputraintan** [comment]
> Just some minor typos that caught my eye in the screenshots @williamputraintan, looks good otherwise!

Thanks Pete!

**Pull request merged**